packages/notifi-node/lib/client/NotifiClient.ts

```diff
@@ -226,6 +223,214 @@ class NotifiClient {
+      targetGroupId: targetGroup.id,
+      filterOptions: '{}',
+    });
+    const alert = createResult.createAlert;
+    if (alert === undefined) {
+      throw new Error('Failed to create alert');
+    }
+
+    return alert;
+  };
+
+  getOrCreateSourceGroup: (
+    name: string,
+  ) => Promise<Gql.SourceGroupFragmentFragment> = async (name) => {
+    const getResult = await this.service.getSourceGroups({});
+    if (getResult.sourceGroup === undefined) {
+      throw new Error('Failed to get SourceGroups');
+    }
+    const existing = getResult.sourceGroup.find((sg) => sg?.name === name);
+    if (existing !== undefined) {
+      return existing;
+    }
+
+    const createResult = await this.service.createSourceGroup({
+      name,
+      sourceIds: [],
+    });
+    if (createResult.createSourceGroup === undefined) {
+      throw new Error('Failed to create SourceGroup');
+    }
+
+    return createResult.createSourceGroup;
+  };
+
+  getOrCreateTargetGroup: (
+    name: string,
+    webhookTarget: Gql.WebhookTargetFragmentFragment,
+  ) => Promise<Gql.TargetGroupFragmentFragment> = async (
+    name,
+    webhookTarget,
+  ) => {
+    const getResult = await this.service.getTargetGroups({});
+    if (getResult.targetGroup === undefined) {
+      throw new Error('Failed to get TargetGroups');
+    }
+
+    const existing = getResult.targetGroup.find((tg) => tg?.name === name);
+    if (existing !== undefined) {
+      return this.updateTargetGroup(existing, webhookTarget);
+    }
+
+    const createResult = await this.service.createTargetGroup({
+      name,
+      emailTargetIds: [],
+      smsTargetIds: [],
+      telegramTargetIds: [],
+      webhookTargetIds: [webhookTarget.id],
+    });
+    if (createResult.createTargetGroup === undefined) {
+      throw new Error('Failed to create TargetGroup');
+    }
+
+    return createResult.createTargetGroup;
+  };
+
+  updateTargetGroup: (
+    targetGroup: Gql.TargetGroupFragmentFragment,
+    webhook: Gql.WebhookTargetFragmentFragment,
+  ) => Promise<Gql.TargetGroupFragmentFragment> = async (
+    targetGroup,
+    webhook,
+  ) => {
+    const updateResult = await this.service.updateTargetGroup({
+      id: targetGroup.id,
+      name: targetGroup.name ?? targetGroup.id,
+      emailTargetIds: [],
+      smsTargetIds: [],
+      telegramTargetIds: [],
+      webhookTargetIds: [webhook.id],
+    });
+
+    const updated = updateResult.updateTargetGroup;
+    if (updated === undefined) {
+      throw new Error('Failed to update targetGroup');
+    }
+
+    return updated;
+  };
+
+  createOrUpdateWebhook: (
+    params: Gql.CreateWebhookTargetMutationVariables,
+  ) => Promise<Gql.WebhookTargetFragmentFragment> = async (params) => {
+    const getResult = await this.service.getWebhookTargets({});
+    const existing = getResult.webhookTarget?.find((w) => {
+      return w.url === params.url && w.format === params.format;
+    });
+
+    if (existing === undefined) {
```

Comment:
I wonder if we should just move to a model where the create functions always return an existing object on attempt to create a dupe. Just to simplify and reduce rtt

Reply:
i think that would be nice

one consideration: the client wouldn't have the `id`, so we would need to move that logic to using `name` as the identifier somehow (we dedupe on `name` inside the SDK).